Add new Grand Rename article explaining why and how it's done

Jeehut opened the pull request:
@winsmith These are my renaming suggestions. If approved, I can create pull requests on all mobile SDKs that apply them. I would also go through other docs and adjust references accordingly once approved, I only adjusted one article so far.

TODOs that still need to be discussed or are out of my scope are:

- [ ] Renaming the SwiftClient to Swift SDK on GitHub. (As long as you don't create a new repo with the old name, GitHub will redirect to the new name automatically.)
- [ ] I marked all the Android lifecycle signals as "Deleted" assuming they would get deleted. Maybe I should even remove them from the list here, but they are currently there, so I added them so we can discuss.

winsmith commented on articles/grand-rename.md:
```diff
@@ -0,0 +1,97 @@
+- `AppInfo`: Information about the specific app build, such as version, build number, or SDKs compiled with.
+- `Device`: All about the device running the application, such as operating system, model name, or architecture.
+- `SDK`: Information about the TelemetryDeck SDK, such as the version number or supported features.
+
```
Add this point please

- `Metrics`: Information about app and device performance

winsmith commented on articles/grand-rename.md:
```diff
@@ -0,0 +1,97 @@
+- `TelemetryDeck.AppContext.language`: The language the app is currently used in (one of the supported languages).
+- `TelemetryDeck.AppInfo.versionAndBuildNumber`: Combines the app version and build into one String, such as `1.7.1 (build 22)`.
+- `TelemetryDeck.Device.preferredLanguage`: The language most preferred by the user on this device (might not be supported by app).
+- `TelemetryDeck.Device.type`: One of `Phone`, `Tablet`, `Desktop`, `Watch`, `TV`, or `Headset`
```
We don't need that information in the SDK, because it can be derived from the modelName. To do that, we use Druid's lookups, and we feed them with [this data](https://github.com/TelemetryDeck/AppleModelNames).

Jeehut replied:
Where do you use Druid's lookups? Can users easily use the device type as well? I can't find any field here:
<img width="531" alt="Bildschirmfoto 2024-04-22 um 17 16 19" src="https://github.com/TelemetryDeck/docs/assets/6942160/8904fb16-e1fe-4f24-a207-4b8127d8043a">

Even if that was possible, I feel like doing such a device type lookup afterwards seems overkill to me and requires regular updates of the library. At least on all Apple platforms, getting the device type is super easy so the SDK could do it. Haven't checked for Android though, maybe it's harder there...

winsmith replied:
Touché, lookups can't be set via the visual editor yet. But I think that should be an easy addition.

winsmith commented on articles/grand-rename.md:
```diff
@@ -0,0 +1,97 @@
+- `TelemetryDeck.Device.screenResolutionHeight`: The resolution height of the screen in pixel/points (whatever is most common on platform).
+- `TelemetryDeck.Device.orientation`: One of `Portrait`, `Landscape`, or `Fixed`.
+- `TelemetryDeck.Device.timeZone`: The timezone expressed by the UTC offset, such as `UTC+0` (London), `UTC+9` (Tokyo), or `UTC-8` (San Francisco).
+- `TelemetryDeck.SDK.supportsPrefixes`: Indicates if the TelemetryDeck SDK version is already migrated to the new prefixed naming system.
```
Can't we derive that form `TelemetryDeck.SDK.version`?

winsmith replied:
Oh and speaking of stuff we can put under the SDK namespace: 

`TelemetryDeck.SDK.version` could be just the version string, with a separate parameter for the SDK ... platform, like `TelemetryDeck.SDK.platform`. Not sure if platform is the correct naming here, do you have better suggestions?

```json
{
  "TelemetryDeck.SDK.version": "2.0.0",
  "TelemetryDeck.SDK.platform": "SwiftSDK"
}
```

Jeehut replied:
I think TelemetryDeck.SDK.platform is quite good, only alternative that comes to my mind is TelemetryDeck.SDK.framework if you want to about platform.

Jeehut replied:
On second thought `platform` doesn't seem quite right given that most SDKs are named by programming languages. If we go by `Platform`, the Kotlin SDK should be renamed to `AndroidSDK` and the Swift one to `AppleSDK`, for example. But I think `platform` is too easy to misunderstand and confuse with `TelemetryDeck.Device.platform`, so we should avoid it.

For `SwiftSDK` / `KotlinSDK` / `FlutterSDK`, we could use `language`, `variant`, `type`, `id`, or simply `name`. I chatted a bit with ChatGPT and we figured `name` is probably the most flexible and clear one. Let me know if you disagree, I don't have a strong feeling about the naming here.

Jeehut replied:
I just split the old `telemetryClientVersion` up to `TelemetryDeck.SDK.name` and `TelemetryDeck.SDK.version`, but for compatibility reasons I kept the old style and renamed that to `TelemetryDeck.SDK.nameAndVersion`. See 1f2f3a5e31f4ced1120e2ab13ed6dfa483b7fe26.

winsmith commented on articles/signal-type-naming.md:
```diff
@@ -31,25 +31,27 @@ What many of our customers do (and we ourselves internally as well) is using a d
 ## Distinguishing between views, actions and events
 
-If you want to distinguish between views, actions and events, or just views and actions, you can add that to the type, such as `InsightEditor.MetaEditor.actions.SaveInsight`. However, I personally don’t get much value from that. Instead, I usually annotate types implicitly using grammar:
+If you want to distinguish between views, actions and events, or just views and actions, you can add that to the type, such as `InsightEditor.MetaEditor.Actions.saveInsight`. However, I personally don’t get much value from that. Instead, I usually annotate types implicitly using grammar:
```
Please use "we" instead of "I"

Jeehut replied:
I will make the adjustment and also do a quick search for other places where "I" is used. If you used "I" here, it might be in other places, too.

winsmith commented on articles/grand-rename.md:
```diff
@@ -0,0 +1,97 @@
+
+{% noteinfo "No action needed" %}
+To make sure these changes don't affect existing customers with existing insights anytime soon, we decided to keep sending the old names for a transition period of at least 1 year while also sending the new ones. We have also auto-migrated any insights that were using the old names in their filters to work with the new system. This was safely achieved by replacing filters like "platform == iOS" w ...
+{% endnoteinfo %}
```
I'd suggest not promising auto-migration right now. Instead, let's promise that existing SDKs (we might call it SwiftSDK 2.0.0) will send new and deprecated parameters, but SwiftSDK 3.0.0 will drop the deprecated parameters when we release it in a year. 

In the coming year, we can then present a sort of "Migration Wizard" that will try to automatically update filters when it runs, and also list GroupBy Queries and similar insights that might need to be manually updated. 

We can either run this Wizard on the server in the background, if we're sure we can auto-migrate everything, or we can run it on the frontend with user interaction if we think some user consent is required.

kkostov replied:
I can imagine some folks may ask if sending both signal types would also mean they'd be billed as two signals?

Not sure how the TD backend works, so I'm thinking why task the client with the migration? Perhaps we could find a way to make it so existing apps don't have to migrate right away but new apps can adopt the new signal types. And then v3 will no longer allow sending deprecated signal types.

For example, in v2.0, we extend the configuration object with a boolean parameter like `TelemetryDeck.useModernSignalTypes` set to `false` by default. This will be included in the additional payload which the server can use to rename the parameters during ingestion.

Users are welcome to set this to `true` in order to adopt the new signal types early during v2 (e.g. for new apps). When v3 is launched, the value is set to true and can no longer be overwritten.

Jeehut replied:
Alright, I will make the adjustment, seems like a good approach as well! 👍

Jeehut replied:
By the way, if you want to make a major new release of the SDK for this rename, I would change my opinion on the scope of this PR and include the consistency of the calls across al SDKs. My main problem with them is that in the docs, we don't have the option to write the correct calls for every SDK, so we have to pick one. But if the APIs all were named consistently, we could simply write one piece of code in the docs, and the other platforms would look very similar.

I just had a great discussion with @kkostov on Discord (we happened to meet today in the channel!) and we agreed that for Swift, Kotlin, and Flutter renaming `TelemetryManager.send(:with:)` to `TelemetryDeck.signal(:parameters:)` and additionally streamlining also the `requestImmediateSync()` approach introduced in https://github.com/TelemetryDeck/SwiftClient/pull/136 (rather than having two separate functions in Android) would be better.

We would also check if the initialization code has differently named APIs and adjust that, too. It should be totally possible to make these renaming changes without breaking existing code (via overloads and maybe deprecations).

What do you think? @winsmith @timidak

Jeehut replied:
@kkostov Oh, I'm just seeing your comment now, it seems we were writing at the same time and GitHub didn't indicate in any way that there was a new comment when I sent mine.

I did think about your approach, but firstly the duplicate parameters won't count as a new signal, so users shouldn't be concerned. The one extra signal we will send for the `TelemetryDeck.Session.started` on app start shouldn't make a big difference as most apps send multiple signals during a session and the percentage of one extra signal per session becomes relatively low.

If anything, I would suggest that we give users the option to opt-out of the deprecated calls and exclusively send the new ones. Users just need to make sure they update their insights to the new ones if they have any in those cases and then they wouldn't have any additional signals. But like I said above, I'm not even sure if this is necessary. If we were to migrate all the Android lifecycle events over, then you would be right, but I'm assuming they are going to get removed soon (maybe alongside the 2.0 migration?). I can imagine they produce a lot of signals, like each time a user looks up something to answer the questions in my crossword puzzle game and and return to the app, it's at least 2 signals for every lookup which can lead to >100 signals per user per puzzle.

Jeehut replied:
@winsmith I adjusted the note as you requested in 8756fd592da067913ea5cbb96b61ce3670448951. I didn't mention any version details, but tried to make clear we are aware of potential migration needs and will do our best to support existing customers in the migration in some way. I think the most important info here is the title: "No immediate action needed".

timidak replied:
Calling it `TelemetryDeck.signal` would be in sync with the JavaScript SDK and our React implementation. So I'd be super okay with that. If I get this correctly `TelemetryDeck.record` would be the same as `TelemetryDeck.queue` in the JavaScriptSDK (cf. [Code in the JsSDK](https://github.com/TelemetryDeck/JavaScriptSDK/blob/1e16779aefe037ddf24a5cef0290d9a6a2bbce12/src/telemetrydeck.js#L75)). 

I personally like the name queue, but I don't have a strong opinion which name might fit better. The other concept sounds to me like the `flush` method in the JavaScript SDK (refer to [this line](https://github.com/TelemetryDeck/JavaScriptSDK/blob/1e16779aefe037ddf24a5cef0290d9a6a2bbce12/src/telemetrydeck.js#L89)).

In general, I don't have a strong opinion, but I can see that naming consistency across SDKs should be addressed.

Jeehut replied:
@timidak What's the difference between `signal` and `queue`? And why aren't you using a verb like `enqueue`?

In Swift, the `signal` method is enqueueing sending the signal by default. There's no API to do it synchronously, so the `signal` is basically an `enqueue` by default. We avoid a synchronous signal to reduce server load. Same for what you call `flush` (is this a term of art? never heard it and on first thought I assumed it deletes all enqueued signals even if not sent yet), to reduce server load I've made sure that it only "directly" sends all enqueued signals if there was no call to "flush" within the last 10 seconds. Therefore the `request` prefix in `requestImmediateSync()`.

But I guess things might be a little different on the web. We should only make consistent, what has a similar behavior from the users perspective. If a synchronous call is the default on Frontend and what people should use, it can make sense to keep the name `signal` which is asynchronous on Mobile. If `flush` also continues to behave differently, it doesn't have to have the same name. But I would suggest `immediateSync` for consistency even if you don't implement the check for "already called in last 10 seconds", which might not be possible in Frontend, I don't know.

winsmith replied:
Background info – Here's how we've been using these words so far, at least to my memory:

- send: Either sends a signal directly, or enqueues is, depending on SDK
- queue, enqueue: Add signal to a queue to be sent later (this is not  to reduce server load, but mostly to reduce antenna activation times for mobile devices)
- flush: send all signals that are in the queue right now.

winsmith replied:
> By the way, if you want to make a major new release of the SDK for this rename, I would change my opinion on the scope of this PR and include the consistency of the calls across all SDKs. My main problem with them is that in the docs, we don't have the option to write the correct calls for every SDK, so we have to pick one. But if the APIs all were named consistently, we could simply write one piece of code in the docs, and the other platforms would look very similar if not exactly the same.

Streamlining is a great idea, let's do this.

winsmith replied:
I'm all for this new nomenclature proposed by @Jeehut:

- Main singletons or namespaces are `TelemetryDeck` not `TelemetryManager`
- The verb for sending a signal is `signal`
- send addition parameters with the `parameters` paramete 
- It's not a payload, its `parameters`

Jeehut replied:
@winsmith If the only reason we were using a queue on mobile devices is to reduce battery drain, then I shouldn't have bothered to write the logic for `requestImmediateSync()` and just made it a simple `syncImmediately()`. I really thought it's about server load. 😅

kkostov replied:
> The verb for sending a signal is signal

So just to be clear @winsmith @Jeehut, the Grand Rename will result in dropping `queue` method and repurposing `send` to provide the same functionality (appending signals to the cache), correct?

Note: In the Kotlin SDK the existing `queue` method drops signals older than 24h. The mindset at the time was that "important, don't drop it" signals would go out using `send` and others using `queue`.

Note2: The Flutter SDK exports a `send` method which routes to `send` on iOS and `queue` on Android.

Jeehut replied:
> So just to be clear @winsmith @Jeehut, the Grand Rename will result in dropping send method and repurposing queue as signal to provide the same functionality (appending signals to the cache), correct?

"Dropping" sounds like an immediate breaking change, but it's correct in general. In the Swift SDK I'm gonna keep all existing methods and just mark them as deprecated while they'll continue to work. But a fix-it is provided (or at least a message) that helps with the migration and no app using the SDK today will run into build errors, but just warnings. Until we decide it's time to fully drop them.

> Note: In the Kotlin SDK the existing queue method drops signals older than 24h. The mindset at the time was that "important, don't drop it" signals would go out using send and others using queue.

I'm not aware of any signal dropping in Swift, I also wouldn't expect that to happen and given that the queue will be the only option going forward, you should probably turn off the "drop after 24 h".

> Note2: The Flutter SDK exports a send method which routes to send on iOS and queue on Android.

That sounds reasonable as the old "send" is similar to "queue" on Android. Both will be "signal" in the future though, and I just created the PR for the future Swift SDK in https://github.com/TelemetryDeck/SwiftClient/pull/153 which contains the renaming changes in commit https://github.com/TelemetryDeck/SwiftClient/pull/153/commits/b6c7d436987bce3f5f8d64c74cf023d925844ba2.

Jeehut replied:
Please note that I found a `for` parameter and also a `floatValue` parameter on the `send` function. I renamed `for` to `customUserID` to be consistent with the configurations `defaultUserIdentifier` (which could be renamed to `defaultUserID` in the future). The `floatValue` didn't change. But I found that the order was like this:

1. Signal Name (required)
2. Custom User ID
3. Float Value
4. Parameters

It's more common to place parameters used more often at the front though and ones that are usually skipped to fall back to the default values at the end. So I changed the order as follows:

1. Signal Name (required)
2. Parameters
3. Float Value
4. Custom User ID

To still provide fix-its for most case, I added a copy of the function which only takes the Signal Name and Parameters and made sure it's favored over the function with more parameters if only 2 are provided. Otherwise, Xcode is unable to apply fix-its when the order of parameters change, but I think fix-its can be really useful to save time in migration.

Jeehut replied:
@kkostov I'm not gonna edit any of the Android / Flutter Setup documents, I think it's safer if you do the changes as soon as you've worked on the rename.

winsmith commented on articles/grand-rename.md:
```diff
@@ -0,0 +1,97 @@
+To make sure these changes don't affect existing customers with existing insights anytime soon, we decided to keep sending the old names for a transition period of at least 1 year while also sending the new ones. We have also auto-migrated any insights that were using the old names in their filters to work with the new system. This was safely achieved by replacing filters like "platform == iOS" w ...
+{% endnoteinfo %}
+
+<!-- Discussion: What about Top-N insights that use a parameter like "appVersion" which got renamed to "TelemetryDeck.SystemInfo.appVersion"? -->
```
I think those by sending both new and deprecated parameter names in parallel for a good long while, we can then auto-update those in the future. (see also previous comment)

Jeehut replied:
Makes sense!

winsmith replied:
@kkostov:
> I can imagine some folks may ask if sending both signal types would also mean they'd be billed as two signals?

Unless I'm misunderstanding @Jeehut, we're only sending duplicate parameters in the same signal. Those don't cost extra.

(I'll try to reply to the rest tomorrow)

Jeehut replied:
Yes, that's what I meant. I do realize though that he has a point that we also have a signal that we're renaming. But it's just one signal, so I'm not sure if it's a big deal for customers. If anything, I would suggest that we allow them to fully migrate to the new system and avoid duplicate calls by skipping the deprecated call. But in any case I would keep the duplicate parameters for both signals.

Jeehut replied:
I removed this comment in 8756fd592da067913ea5cbb96b61ce3670448951 and I agree with you that because we're sending the old names in parallel rather than doing the auto-migration, this is a non-issue at this moment.

winsmith reviewed (commented):
I like these a lot, individual thoughts follow.

- This is indeed the perfect time to also rename the Swift Client to Swift SDK.
- `operatingSystem` and `platform` – I think there is some history here, but 
- Regarding these namespaced entries, we also have `TelemetryDeck.Presets.Onboarding.step`. Should `Presets` also be a namespace, or do you want to rename that to something else? 
- Can you please note which entries should not be strings? If we're sending floats or ints (which is very advisable in some cases) , I'll need to tell the server in advance.
- I very much like the idea of running old and new signal parameters in parallel for a year. We can then also warn users to migrate off older versions of the SDK using a banner or something.

winsmith commented:
Hey @kkostov, @pichfl and @timidak, I'd love your feedback on these changes regarding renaming and cleaning up parameters the SDK is sending.

Jeehut commented:
@winsmith I just addressed all comments and also updated my original comment here, removing the TODOs we already clarified and adding one new entry. I also updated https://github.com/TelemetryDeck/SwiftClient/pull/149 accordingly with the suggested changes. Now to your main comment points:

> operatingSystem and platform – I think there is some history here

Does that mean we should keep them "as-is"? I can understand that, I have no strong feelings here. Just wanted to bring it up that it confused me while we were talking renaming stuff.

> Regarding these namespaced entries, we also have TelemetryDeck.Presets.Onboarding.step. Should Presets also be a namespace, or do you want to rename that to something else?

I think we should stick to 3 parts, so `TelemetryDeck.Presets.Onboarding.step` is too long. I will make sure to adjust the PRs where I introduce the presets after this has been merged (or maybe before). But I think it's a separate discussion. I will add any new group names to the renaming document on those PRs.

> Can you please note which entries should not be strings? If we're sending floats or ints (which is very advisable in some cases) , I'll need to tell the server in advance.

For these two new fields sending a number could help:

1. `TelemetryDeck.Device.screenResolutionWidth`
2. `TelemetryDeck.Device.screenResolutionHeight`

Let me know if you change them, I have implemented them as Strings for now.
There's one more we might send an integer for, but I'm not sure if it really makes a difference:

1. `TelemetryDeck.AppInfo.buildNumber`: This is an integer on all mobile platforms.

Jeehut commented:
I want to point out a few changes I made in the latest commits:

1. I renamed `AppContext` to `RunContext` to further clarify its difference to `AppInfo` to avoid confusion.
2. I introduced the new prefix `UserPreference` for `region` and `preferredLanguage`, the latter I could then shorten to just `language`.

I would like to also point out one renaming decision I had made earlier in this PR which is easy to miss:
I had changed the order of words in the old `majorSystemVersion` and `majorMinorSystemVersion` in the new names to have the same prefix `system` (for the very same reasons we have prefixes at all: better grouping). They are now `systemVersion`, `systemMajorVersion`, and `systemMajorMinorVersion`.

Jeehut commented:
I just updated more documents alongside creating https://github.com/TelemetryDeck/SwiftClient/pull/153 which applies the changes to the SwiftSDK. Please note that before we ship this to production, I should be doing a full test with the whole onboarding process from creating a new app to setting up everything following the docs to make sure nothing is broken.

I did not touch the Objective-C code by the way, neither in the docs, nor in the SDK. I consider Obj-C as "deprecated" in itself, so renaming anything there seems to be waste of time to me. Let me know if you think otherwise.

winsmith commented:
> I just updated more documents alongside creating [TelemetryDeck/SwiftClient#153](https://github.com/TelemetryDeck/SwiftClient/pull/153) which applies the changes to the SwiftSDK. Please note that before we ship this to production, I should be doing a full test with the whole onboarding process from creating a new app to setting up everything following the docs to make sure nothing is broken.

Fantastic work, and yes please do a full test run.

> I did not touch the Objective-C code by the way, neither in the docs, nor in the SDK. I consider Obj-C as "deprecated" in itself, so renaming anything there seems to be waste of time to me. Let me know if you think otherwise.

Sensible, I agree with this decision.

Jeehut commented:
@winsmith I just did a full setup run to test if the Swift setup guide looks good and everything is right with the Swift SDK. I found a couple of bugs and fixed them. From my point of view, the Swift part of the grand rename is now ready. So this and https://github.com/TelemetryDeck/SwiftClient/pull/153 could be both merged. Don’t forget to rename the SwiftClient to SwiftSDK after the merge.

I personally don’t think there’s need to wait for the Android/Flutter SDK before merging this, because:
1. I have not touched any of their setup guides (only the Swift one), so they can be updated at a later time.
2. As long as we don’t market the “Grand Rename” article, no one will be aware of it.
3. As long as there’s no new 2.0.0 release on GitHub of the SwiftSDK, no one will notice the change (unless they rely on the `main` branch, which should be rare).

But feel free to take your time to think about it and give me feedback.

I’ll be focusing on the Purchases & Errors presets feature next. I will rebase the branches on top of the grand-rename branches and make sure make all adjustments needed and to include a built-in solution for the presets in the SwiftSDK so setting up the presets is even less typing work.
